libtenzir/builtins/operators/throttle.cpp

```diff
@@ -0,0 +1,162 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/detail/alarm_clock.hpp>
+#include <tenzir/diagnostics.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/tql/fwd.hpp>
+#include <tenzir/tql/parser.hpp>
+#include <tenzir/tql2/plugin.hpp>
+
+#include <algorithm>
+#include <ranges>
+
+namespace tenzir::plugins::throttle {
+
+namespace {
+
+using float_seconds = std::chrono::duration<double>;
+
+auto split_chunk(const chunk_ptr& in,
+                 size_t position) -> std::pair<chunk_ptr, chunk_ptr> {
+  if (position >= in->size()) {
+    return {in, chunk::make_empty()};
+  }
+  return {in->slice(0, position), in->slice(position)};
+}
+
+} // namespace
+
+class throttle_operator final : public crtp_operator<throttle_operator> {
+public:
+  throttle_operator() = default;
+
+  explicit throttle_operator(double max_bandwidth, float_seconds window)
+    : bandwidth_per_second_(max_bandwidth / window.count()), window_(window) {
+  }
+
+  // TODO: Currently the operator only handles byte stream, but in the future
+  // we also want to be able to handle events as input.
+  auto operator()(generator<chunk_ptr> input,
+                  operator_control_plane& ctrl) const -> generator<chunk_ptr> {
+    auto alarm_clock = ctrl.self().spawn(detail::make_alarm_clock);
+    auto last_timestamp = std::chrono::system_clock::now();
```

Comment:
Throughout this function, you're only ever looking at time deltas, so I think it'd be fine to use a steady clock instead?